src/utils.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""A collection of utility functions."""
+
+from os import readlink
+from pathlib import Path
+
+
+def find_packages_by_indices(indices, base=""):
+    """Return the location of the packages.
+
+    Find the locations of the packages given the path to "Packages*" files.
+
+    Args:
+        indices: A list of index files.
+        base: The path prepend to the returned filename (optional, default="").
+
+    Returns:
+        The an unique set of path-to-packages found in the indices.
+    """
+    packages = set()
+    package_indices = indices if isinstance(indices, list) else [indices]
+    for package_index in package_indices:
+        packages |= _locate_packages_from_index(package_index, archive_root=base)
+    return packages
+
+
+def locate_package_indices(path):
+    """Return all locations of the index files.
+
+    Find the locations of the index files within `path`.
+
+    Args:
+        path: The path from which we search for the index files.
+
+    Returns:
+        A zip of the path to archive root and the path to the package
+        indices.
+    """
+    path = Path(path)
+    if not path.exists():
+        raise FileNotFoundError("Invalid path: {}".format(path))
+
+    dists = list(path.glob("**/dists"))
+    pool = [d.parent / "pool" for d in dists]
+
+    archive_roots = []
+    package_indices = []
+    for p, d in zip(pool, dists):
+        archive_roots.append(
+            Path(readlink(p)).parent.absolute()
+            if p.is_symlink()
+            else p.parent.absolute()
+        )
+        package_indices.append(
+            [sorted(distro.glob("**/Packages*"))[0] for distro in d.glob("*")]
+        )
+    return zip(archive_roots, package_indices)
+
+
+def convert_bytes(num):
+    """Convert `num` of bytes to an appropriate unit."""
+    for unit in ["bytes", "KB", "MB", "GB", "TB"]:
+        if num < 1024:
+            return "{:.1f} {}".format(num, unit)
+        num /= 1024
+
+
+def _locate_packages_from_index(package_index, archive_root=""):
+    """Return an unique set of path-to-packages found in the index file."""
+    opener = _get_opener(package_index)
+    with opener(package_index, mode="r") as f:
+        packages = {
+            Path(archive_root, line.strip().replace("Filename: ", ""))
+            for line in f
+            if line.startswith("Filename: ")
+        }
+    return packages
+
+
+def _get_opener(path):
+    """Return appropriate opener to open an index file."""
+    extension = Path(path).suffix
+    if extension == ".gz":
+        from gzip import open as gopen
```

Comment:
Is there a chance that the system wont have all three of these packages (`bz2`/`gzip`/`lzma`)? If not I think they should be imported at the top of the module.

Reply:
it is quite likely we don't have all three of them, this is mainly use for opening the "Package" file under `**/dists/**/Package*`. That file can be in any of these formats as mentioned in the debian package format documentation, and it up to the maintainer to choose which or all formats are in that directory.

Reply:
but shouldn't the charm install all of them to be prepared to deal with whatever maintainer chooses? (both compression programs and their python bindings)

Reply:
Just to be more clear, my original question was more about the conditional imports. Why aren't we importing packages at top of the? Is there a chance that the system won't have some of these (python) packages available?

Bottom line is that I think that these import should be on top 😆

Reply:
Oh, I miss understood your question.. Yeah, I believe they exist in Python 3s' standard libaries. The only reason that I don't do that is that it's very likely that we don't use them all. But after a second thought, these modules is pretty small, it probably doesn't hurt to import them at the top.